src/DoctrineBatchUtils/BatchProcessing/SelectBatchIteratorAggregate.php

```diff
@@ -0,0 +1,138 @@
+ * 'Read' focused batching iterator that will issue a clear to the entity manager
+ * every batch size to 'detach' the managed objects and therefore make them entities only for the
+ * purpose of reading without manually attaching them back to the EntityManager.
+ */
+final class SelectBatchIteratorAggregate implements IteratorAggregate
+{
+    /**
+     * @var Traversable
+     */
+    private $resultSet;
+
+    /**
+     * @var EntityManagerInterface
+     */
+    private $entityManager;
+
+    /**
+     * @var int
+     */
+    private $batchSize;
+
+    public static function fromQuery(AbstractQuery $query, int $batchSize): self
+    {
+        return new self($query->iterate(), $query->getEntityManager(), $batchSize);
+    }
+
+    /**
+     * @param object[] $results
+     * @param EntityManagerInterface $entityManager
+     * @param int $batchSize
+     *
+     * @return self
+     */
+    public static function fromArrayResult(
+        array $results,
+        EntityManagerInterface $entityManager,
+        int $batchSize
+    ): self {
+        return new self(new ArrayIterator($results), $entityManager, $batchSize);
+    }
+
+    public static function fromTraversableResult(
+        Traversable $results,
+        EntityManagerInterface $entityManager,
+        int $batchSize
+    ): self {
+        return new self($results, $entityManager, $batchSize);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getIterator()
+    {
+        $iteration = 0;
+        $resultSet = $this->resultSet;
+
+        foreach ($resultSet as $key => $value) {
+            $iteration += 1;
+
+            if (is_array($value)) {
+                $firstKey = key($value);
+                if ($firstKey !== null && is_object($value[$firstKey]) && [$firstKey => $value[$firstKey]] === $value) {
+                    yield $key => $this->reFetchObject($value[$firstKey]);
+
+                    $this->clearBatch($iteration);
+                    continue;
+                }
+            }
+
+            if (! is_object($value)) {
+                yield $key => $value;
+
+                $this->clearBatch($iteration);
+                continue;
+            }
+
+            yield $key => $this->reFetchObject($value);
+
+            $this->clearBatch($iteration);
+        }
+
+        $this->entityManager->clear();
+    }
+
+    /**
+     * BatchIteratorAggregate constructor (private by design: use a named constructor instead).
+     *
+     * @param Traversable $resultSet
+     * @param EntityManagerInterface $entityManager
+     * @param int $batchSize
+     */
+    private function __construct(Traversable $resultSet, EntityManagerInterface $entityManager, int $batchSize)
+    {
+        $this->resultSet     = $resultSet;
+        $this->entityManager = $entityManager;
+        $this->batchSize     = $batchSize;
+    }
+
+    /**
+     * @param object $object
+     *
+     * @return object
+     */
+    private function reFetchObject($object)
```

Comment:
`object` can be used as a hint

Reply:
It can't because this package is 7.1

Reply:
Bump to 7.2

Reply:
Id prefer not to given my project is 7.1, also given its not a breaking change there isn't a real need for a major revision. There is still 20~ days left that 7.1 is feature supported and another year for security.

If thats ok Id prefer we kept it, however I will make another PR after this is merged that is bumped to 7.2 and I can update the other concrete

Reply:
I'm always targeting the latest stable for minor releases, regardless.

> There is still 20~ days left that 7.1 is feature supported and another year for security.

Doesn't affect the package.

 > If thats ok Id prefer we kept it, however I will make another PR after this is merged that is bumped to 7.2 and I can update the other concrete

This branch already targets `master`, which will be `2.0.0` anyway. Bumping the dependency is fine. BTW, bumping in minor is also OK, it's just imperative not to bump versions in patch releases (since security fixes must be installable on all supported systems).

Reply:
Okay I will bump and type hint, then fork a version for now until I can move the project onto 7.2, its on our roadmap but wont be this year.